File: sol/1942/1942.cpp
struct Solution {
    int smallestChair(vector<vector<int>>& times, int targetFriend) {
        int n = times.size();
        vector<int> idx(n);
        ranges::iota(idx, 0);
        ranges::sort(idx, [&](int i, int j){return times[i][0] < times[j][0];});
        int nextChair = 0;
        vector<pair<int, int>> peopleHeap;
        vector<int> chairHeap;
        for (int i : idx){
            int chair = -1;
            while (!peopleHeap.empty() && -peopleHeap.front().first <= times[i][0]){
                ranges::pop_heap(peopleHeap);
                chairHeap.push_back(peopleHeap.back().second);
                ranges::push_heap(chairHeap);
                peopleHeap.pop_back();
            }
            if (chairHeap.empty()){
                chair = (nextChair++);
            } else {
                ranges::pop_heap(chairHeap);
                chair = -chairHeap.back();
                chairHeap.pop_back();
            }
            peopleHeap.emplace_back(-times[i][1], -chair);
            ranges::push_heap(peopleHeap);
            if (i == targetFriend){
                return chair;
            }
        }
        return -1;
    }
};

File: sol/1942/1942.py
import heapq as hq

class Solution:
    def smallestChair(self, times: List[List[int]], targetFriend: int) -> int:
        n = len(times)
        idx = list(range(n))
        idx.sort(key=lambda i: times[i][0])
        nextChair = 0
        peopleHeap = []
        chairHeap = []
        for i in idx:
            chair = -1
            while len(peopleHeap) != 0 and peopleHeap[0][0] <= times[i][0]:
                _, emptyChair = hq.heappop(peopleHeap)
                hq.heappush(chairHeap, emptyChair)
            if len(chairHeap) == 0:
                chair = nextChair
                nextChair += 1
            else:
                chair = hq.heappop(chairHeap)
            hq.heappush(peopleHeap, (times[i][1], chair))
            if i == targetFriend:
                return chair
        return -1
